src/matrix_cookbook/9_special_matrices.lean

```diff
@@ -39,41 +45,313 @@ det_from_blocks₁₁ _ _ _ _
+  },
+  { -- A12 Block
+    rw [matrix.add_mul, add_assoc], 
+    repeat {rw matrix.neg_mul}, 
+    rw [←sub_eq_add_neg, matrix.mul_assoc _ _ (⅟C₂), matrix.mul_assoc (⅟A₁₁⬝(A₁₂⬝⅟C₂)), 
+      matrix.mul_assoc, ←matrix.mul_sub (⅟A₁₁⬝(A₁₂⬝⅟C₂)), matrix.mul_assoc, 
+      ← neg_sub, matrix.mul_neg, matrix.mul_inv_of_mul_self_cancel],
+    rw [matrix.mul_neg, add_right_neg],
+  },
+  { -- A21 Block
+    rw [matrix.mul_assoc, matrix.inv_of_mul_self, matrix.mul_one, matrix.neg_mul, neg_add_self],
+  },
+  { -- A22 Block
+    rw [matrix.mul_assoc, matrix.mul_assoc, matrix.neg_mul, ← matrix.mul_neg,
+    ← matrix.mul_add, neg_add_eq_sub, ← matrix.mul_assoc, matrix.inv_of_mul_self],
+  }
+end
 
 /-! ### Block diagonal -/
 
-lemma eq_401 (A₁₁ : matrix m m R) (A₂₂ : matrix n n R) :
-  (from_blocks A₁₁ 0 0 A₂₂)⁻¹ = from_blocks A₁₁⁻¹ 0 0 A₂₂⁻¹ := sorry
+lemma eq_401 (A₁₁ : matrix m m R) (A₂₂ : matrix n n R)
+  [h1: invertible A₁₁] [h2: invertible (A₂₂)] :
+  (from_blocks A₁₁ 0 0 A₂₂)⁻¹ = from_blocks A₁₁⁻¹ 0 0 A₂₂⁻¹ := 
+begin
+  apply inv_eq_right_inv,
+  rw from_blocks_multiply,
+  simp only [mul_inv_of_invertible, matrix.zero_mul, 
+  add_zero, matrix.mul_zero, zero_add, from_blocks_one],
+end
 lemma eq_402 (A₁₁ : matrix m m R) (A₂₂ : matrix n n R) :
   det (from_blocks A₁₁ 0 0 A₂₂) = det A₁₁ * det A₂₂ := det_from_blocks_zero₁₂ _ _ _
 
 /-! ### Schur complement -/
 
 /-! ## Discrete Fourier Transform Matrix, The -/
 
-lemma eq_403 : sorry := sorry
-lemma eq_404 : sorry := sorry
-lemma eq_405 : sorry := sorry
-lemma eq_406 : sorry := sorry
-lemma eq_407 : sorry := sorry
-lemma eq_408 : sorry := sorry
-lemma eq_409 : sorry := sorry
-lemma eq_410 : sorry := sorry
-lemma eq_411 : sorry := sorry
+/- lemma eq_403 : sorry := sorry
+-- Perhaps we can write eq 403 as a definition -/
+
+
+section dft_matrices
+open_locale real matrix big_operators
+open matrix
+open equiv equiv.perm finset
+open complex
+open polynomial
+
+lemma vec_eq_vec_iff (v w: n → R) : 
+  v = w ↔ ∀ k : n, v k = w k := 
+begin
+  split,
+  intros h k, rw h,
+
+  intros h, ext k,
+  specialize h k, exact h,
+end
+
+lemma mat_eq_mat_iff (v w: matrix m n R) : 
+  v = w ↔ ∀ (j: m)(k: n), v j k = w j k:= 
+begin
+  split,
+  rintros h i j, rw h at *,
+  intro h,
+  ext i j, specialize h i j, exact h,
+end
+
+-- variable {N: ℕ}
+/- Are we really forced to have the DFT matrix 
+noncomputable because the complex exponnetial function is 
+non-computable?
+-/
+
+-- eq_403 : Twiddle Factors
+noncomputable def Wkn {N: ℕ} (k n: fin N) : ℂ :=  
+complex.exp(complex.I * 2 * π * k * n / N)
+
+-- Forward DFT Matrix
+noncomputable def W_N {N: ℕ}: matrix (fin N) (fin N) ℂ :=
+λ (k n: fin N), Wkn k n
+
+-- Inverse Twiddle Factors
+noncomputable def iWkn {N: ℕ} (k n: fin N) : ℂ :=  
+complex.exp(-complex.I * 2 * π * k * n / N)
+
+-- Inverse DFT Matrix
+noncomputable def iW_N {N: ℕ} : matrix (fin N) (fin N) ℂ :=
+λ (k n: fin N), iWkn k n
```

Comment:
I would suggest merging these; there's no need to have them as separate definitions